nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp

```diff
@@ -177,6 +177,10 @@ class BtActionNode : public BT::ActionNodeBase
       on_new_goal_received();
     }
 
+    if (!goal_handle_) {
+      throw std::logic_error("BtActionNode::Tick: invalid goal handle");
```

Comment:
Sure.... that would make it exit in the condition that the `goal_handle_` is already complete, but lets think about this more strategically -- because otherwise your navigation system every time you run into this issue is going to be throwing this error all over the place. 

What situations do you have that goal_handle_ is not valid? Answer that main question first. Is it only when canceled or aborted? If so, then in the result callback, you can check if its not valid and then have an else statement to set the result type to that return code https://github.com/ros-planning/navigation2/blob/7877a1225c549aa0ced78a0f263cbff31829761c/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L276-L279. Then since `result_` has a populated value, you get a clean exit on the next iteration. 

Rather than masking this issue, we should work to fix it. When / why is `goal_handle_` being turned into a nullptr? Answer that, and we can find a real solution -- whether that's fixing an issue in this file, or in the file that's nulling out the pointer

Reply:
Ok. I will start investigation.

Reply:
```
    send_goal_options.result_callback =
      [this](const typename rclcpp_action::ClientGoalHandle<ActionT>::WrappedResult & result) {
        // TODO(#1652): a work around until rcl_action interface is updated
        // if goal ids are not matched, the older goal call this callback so ignore the result
        // if matched, it must be processed (including aborted)
        if (!this->goal_handle_) {
          RCLCPP_ERROR(node_->get_logger(), "error id: %d", result.code);
        }
        if (this->goal_handle_ && (this->goal_handle_->get_goal_id() == result.goal_id)) {
          goal_result_available_ = true;
          result_ = result;
        }
      };
```

```
[bt_navigator-11] [ERROR] [1618637514.271798607] [bt_navigator_rclcpp_node]: error id: 4
[bt_navigator-11] [ERROR] [1618637514.272602967] [bt_navigator_rclcpp_node]: error id: 5
[bt_navigator-11] [ERROR] [1618637647.774230873] [bt_navigator_rclcpp_node]: error id: 4
[bt_navigator-11] [ERROR] [1618637647.839067741] [bt_navigator_rclcpp_node]: error id: 4
[bt_navigator-11] [ERROR] [1618637698.121035886] [bt_navigator_rclcpp_node]: error id: 5
[bt_navigator-11] [ERROR] [1618637712.920679316] [bt_navigator_rclcpp_node]: error id: 5
```

According to experiment, `result.code` is 4 or 5(namely SUCCEEDED or CANCELED).

Reply:
My elementary question is that how `goal_handle_` switches the nullptr to a valid handle.

As far as I can see, when the BT node is firstly ticked, `goal_handle_` is still nullptr.

There is only one the assignment of `goal_handle_`:
https://github.com/ros-planning/navigation2/blob/7877a1225c549aa0ced78a0f263cbff31829761c/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L293
.

So if `goal_handle_` is valid, `on_new_goal_received` was called.

When `on_new_goal_received` is firstly called, `goal_handle_` is still nullptr.

Reply:
> What situations do you have that goal_handle_ is not valid? 

It is when `on_new_goal_received` is not called yet.

> When / why is goal_handle_ being turned into a nullptr?

As far as I can see, `goal_handle_` is not turned into a nullptr.
`goal_handle_` is originally a nullptr when BT node is created.
But original code maked assumption that `goal_handle_` is always valid.

Reply:
@tzskp1 https://github.com/ros-planning/navigation2/blob/d7beb7104b952f89399fac7b78b80074a7d55ddc/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L177 `on_new_goal_received` is the first function called after `on_tick()` which populates that `goal_` on the first iteration. 

It is incorrect that we always assume its valid, but we need to then reset this handle properly when its terminated and also know what a nullpointer means in the context

Reply:
> its another function, not code called on the first call of `on_new_goal_received`

As far as I know, this spin calls the callback.
https://github.com/ros-planning/navigation2/blob/7877a1225c549aa0ced78a0f263cbff31829761c/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L284
So the assignment statement follows the callback.

My intention is that nullptr should be one of the following.
- BT node is created
- BT node is halted